### PR TITLE
Bug 1340344 - Add additional interfaces implementation for handling watches, rsh, etc.

### PR DESCRIPTION
**Pull request by soltysh**
Fixes [bug1340344](https://bugzilla.redhat.com/show_bug.cgi?id=1340344). 

@danmcp @mfojtik ptal

### CODE REVIEW

**Comment by soltysh** [comment]
Without the fix everything from watches to rsh, exec, attach is broken.

[test]

**Comment by liggitt** [comment]
I didn't realize we were wrapping the request for audit. I'm a little concerned with that approach, for reasons like this. Is there a reason we can't just log the info we already have from admission?

**Comment by soltysh** [comment]
@liggitt the problem is we need both the request and response information. Without the latter, I wouldn't have to wrap the reponsewriter.

**Comment by liggitt** [comment]
Is that just for the response code? I'm not sure I'd want the API server response handling modified just to accomplish that.

**Comment by soltysh** [comment]
Correct, that's there only the get that response code.

File: pkg/cmd/server/origin/audit.go [hunk 1]
```diff
@@ -52,7 +80,20 @@ func (c *MasterConfig) auditHandler(handler http.Handler) http.Handler {
...
 }
+
+func constructResponseWriter(responseWriter http.ResponseWriter, id string) http.ResponseWriter {
+	delegate := &auditResponseWriter{ResponseWriter: responseWriter, id: id}
+	// check if the ResponseWriter we're wrapping is the fancy one we need
+	// or if the basic is sufficient
+	_, cn := responseWriter.(http.CloseNotifier)
+	_, fl := responseWriter.(http.Flusher)
+	_, hj := responseWriter.(http.Hijacker)
+	if cn && fl && hj {
```
Comment (mfojtik) [inline pkg/cmd/server/origin/audit.go]: can we have a quick unit test to prove this works for attach?

**Comment by soltysh** [comment]
@mfojtik test added, ptal

**Comment by mfojtik** [comment]
lgtm

**Comment by soltysh** [comment]
@liggitt since you where the only one in opposition to this, are you ok with merging this?

**Comment by liggitt** [comment]
The more invasive auditing becomes, the more tests we need around it. If we're wrapping responsewriters, we need tests for fancy requests (watch, exec/attach, log, proxy) with audit enabled

**Comment by soltysh** [comment]
@liggitt I've added unit test per @mfojtik question and basic integration that checks if watch works. I'll be working on more thorough coverage for that feature as a followup, but I'd like to get this in asap to cherry-pick if for ose.

**Comment by mfojtik** [comment]
still LGTM, @liggitt ?

**Comment by liggitt** [comment]
it's ok as a stop-gap, as long as exec/proxy/log was verified manually (wrapping if all interface types are detected is no less safe than what we're doing today), but it seems like the all-or-nothing interface detection approach for the fancy wrapper could bite us in the future... have you swept to make sure all the implementations that satisfy one of those satisfy all? how will we catch drift in the future? (a new impl shows up that only implements one of those interfaces, so it doesn't get wrapped, yet code paths still count on being able to use that one interface)

**Comment by soltysh** [comment]
@liggitt I'm working on more tests to cover all current possible cases. For the new ones, unfortunately you're right it might be problematic, but again it'll be discovered at the time of introducing those features and should be covered by tests cases then. Based on the almost lgtm from Jordan

[merge]

**Comment by soltysh** [comment]
@brenton we need this to be picked up for 3.2.1 and 3.2.0.z as well, once it merges in origin.

**Comment by soltysh** [comment]
[merge[

**Comment by mfojtik** [comment]
[merge]

(should merge now)

**Pull request merged**